### PR TITLE
[firewall rules] Fall back to any host or any protocol when filter list is empty

### PR DESCRIPTION
**Pull request by david-crespo**
The diff is only a couple of lines with [whitespace changes hidden](https://github.com/oxidecomputer/omicron/pull/3511/files?w=1).

This is one possible approach to addressing #3373. Ports already worked this way: `None` or empty vec cause the filter to match any port. But for hosts and and protocols, only `None` falls back to any, while an empty vec will get passed through and presumably never match anything.

I think even if we end up fixing the API types so that an empty `hosts` array is rejected with a 400, we still probably want this logic to be fixed here.

### CODE REVIEW

**Comment by david-crespo** [comment]
Hey @plotnick (because you wrote this) and @internet-diglett (because you’re working on networking APIs now): do you think it makes sense to make ports, hosts, and protocols all treat empty vec the same as null, i.e., matching anything? Without this change, we only do that for ports. I am probably going to tweak the external API to disallow empty vec anyway, but consistency here seems reasonable.

File: illumos-utils/src/opte/firewall_rules.rs [hunk 1]
```diff
@@ -61,41 +61,39 @@ impl FromVpcFirewallRule for VpcFirewallRule {
...
-                        }
-                        HostIdentifier::Vpc(vni) => {
-                            Address::Vni(Vni::new(u32::from(*vni)).unwrap())
-                        }
-                    })
-                    .collect::<Vec<Address>>()
-            },
-        )
+        match self.filter_hosts {
+            Some(ref hosts) if hosts.len() > 0 => hosts
```
Comment (plotnick) [inline illumos-utils/src/opte/firewall_rules.rs:65]: The copy of clippy running in my brain saw this and suggested `!hosts.is_empty()`.

**Review by plotnick — COMMENTED** [review]
So, I think interpreting the empty vec as `Any` for all of hosts, ports, and protocols makes sense, because a rule that never matches anything is pretty useless. It's also how OPTE interprets its filters.

That said, it might not be obvious from the user's perspective, so we should be sure that the behavior is documented and probably indicated somehow in the console.

From a programming perspective, I think if we're going to go this route, we should always use `Vec` instead of `Option<Vec>` in all of the relevant interfaces, but that definitely doesn't have to be done now.

**Comment by luqmana** [comment]
Thanks for taking a look at this @david-crespo! For an immediate fix this works and is fairly small. I feel like I've gone back and forth a bit on whether `Any` should just be an explicit option in the API itself.

**Comment by luqmana** [comment]
@askfongjojo if we're fine with leaving things a bit implicit I think we can mark #3373 as R1. Making things more explicit in the API would be a bit more substantial and require more changes (in omicron and console + sdk/cli too I imagine).

**Comment by david-crespo** [comment]
Cool. I'll merge this, then. 

I'm working on something for the API side that doesn't seem too invasive — I don't think it would require much of a CLI change, if any. The console change would be a couple of lines and I already know which lines. In any case, if I can get it working, I'll push it up and we can decide whether it's worth it.

**Comment by askfongjojo** [comment]
> @askfongjojo if we're fine with leaving things a bit implicit I think we can mark #3373 as R1. Making things more explicit in the API would be a bit more substantial and require more changes (in omicron and console + sdk/cli too I imagine).

Yes, I am fine with the solution. Ship it.

**Pull request merged**